File: Root-finding/Matlab/secant.m
function [x, nsteps] = secant (f, a, b, nmax, tol)
% This function uses secant method to find the root of an equation.
% [x, n] = secant (@(x) x^2-3, 1, 2, 10, 0.01)

X = [a b];

if nargin == 3
    nmax = 100;
    tol = 0.01;
end

for n=1:nmax
    X(n+2) = (X(n) * f(X(n+1)) - X(n+1) * f(X(n))) / (f(X(n+1)) - f(X(n)));
    c = abs(f(X(n+2)));
    
    if c < tol
        break 
    end
end

x = X(end);
nsteps = n;

File: Root-finding/R/secant.R
f <- function(x) x^2-3

secant <- function(f,a,b,nmax=100, tol=0.01){
  # This function uses secant method to find the root of an equation.
  # result = secant (f, 1, 2,50, 0.01)
  X = {}
  X = c(a, b)
  
  for (n in seq.int(1,nmax)){
    temp = (X[n] * f(X[n+1]) - X[n+1] * f(X[n])) / (f(X[n+1]) - f(X[n]))
    X = c(X, temp)
    
    c = abs(f(X[n+2]))
    
    if (c<tol){
      break
    }
    
  }
  
  x = X[n+2]
}